code/modules/xenoarchaeology/xenoartifact_labeler.dm

```diff
@@ -0,0 +1,247 @@
+		return
+
+	trait_toggle(action, "activator", activator_traits, activator)
+	trait_toggle(action, "minor", minor_traits, minor_trait)
+	trait_toggle(action, "major", major_traits, major_trait)
+	trait_toggle(action, "malfunction", malfunction_list, malfunction)
+
+	. = TRUE
+	update_icon()
+
+/obj/item/xenoartifact_labeler/proc/get_trait_list_desc(list/traits, trait_type)//Get a list of all the specified trait types names, actually
+	trait_type = typesof(trait_type)
+	for(var/t in trait_type)
+		var/datum/xenoartifact_trait/X = t
+		if(initial(X.desc) && !(initial(X.desc) in traits) && !(initial(X.label_name)))
+			traits += initial(X.desc)
+		else if(initial(X.label_name) && !(initial(X.label_name) in traits)) //For cases where the trait doesn't have a desc or is tool cool to use one
+			traits += initial(X.label_name)
+	return traits
+
+/obj/item/xenoartifact_labeler/proc/look_for(list/place, culprit) //This isn't really needed but, It's easier to use as a function. What does this even do?
+	for(var/X in place) //Using locate breaks this. Sorry.
+		if(X == culprit)
+			. = TRUE
+	return
+
+/obj/item/xenoartifact_labeler/afterattack(atom/target, mob/user)
+	..()
+	var/obj/item/xenoartifact_label/P = create_label(sticker_name)
+	if(!P.afterattack(target, user, TRUE)) //In the circumstance the sticker fails, usually means you're doing something you shouldn't be
+		qdel(P)
+
+///reset all the options	
+/obj/item/xenoartifact_labeler/proc/clear_selection()
+	sticker_name = null
+	info_list = list()
+	sticker_traits = list()
+	activator = list()
+	minor_trait = list()
+	major_trait = list()
+	malfunction = list()
+	ui_update()
+
+/obj/item/xenoartifact_labeler/proc/create_label(new_name)
+	var/obj/item/xenoartifact_label/P = new(get_turf(src))
+	if(new_name)
+		P.name = new_name
+		P.set_name = TRUE
+	P.trait_list = sticker_traits
+	P.info = activator+minor_trait+major_trait+malfunction
+	return P
+
+/obj/item/xenoartifact_labeler/proc/trait_toggle(action, toggle_type, var/list/trait_list, var/list/active_trait_list)
+	var/datum/xenoartifact_trait/description_holder
+	var/new_trait
+	for(var/t in trait_list)
+		new_trait = desc2datum(t)
+		description_holder = new_trait
+		if(action == "assign_[toggle_type]_[t]")
+			if(!look_for(active_trait_list, t))
+				active_trait_list += t
+				info_list += initial(description_holder.label_desc)
+				sticker_traits += new_trait
+			else
+				active_trait_list -= t
+				info_list -= initial(description_holder.label_desc)
+				sticker_traits -= new_trait
+
+/obj/item/xenoartifact_labeler/proc/desc2datum(udesc) //This is just a hacky way of getting the info from a datum using its desc becuase I wrote this last and it's not heartbreaking
+	for(var/t in typesof(/datum/xenoartifact_trait))
+		var/datum/xenoartifact_trait/X = t
+		if((udesc == initial(X.desc))||(udesc == initial(X.label_name)))
+			return t
+	return "[udesc]: There's no known information on [udesc]!."
```

Comment:
You shouldn't return a string if the function returns a datum. Return null instead or throw an error.